### ISSUE TITLE
What's the idiomatic way to express f(x) = f1(x) + f2(x), and get f's gradient?

### ISSUE THREAD

**Issue by currymj** [kul-forbes/ProximalOperators.jl#49]
I'm assuming f1 and f2 both have gradients.

`SlicedSeparableSum` will compute the function (just provide `(f1, f2)` as the first arguments, and each slice consists of the whole vector). However, it's unable to compute the gradient (which I'm assuming is reasonable in the most general case of slices, even if all input functions have gradients?).

By looking at some of the examples I also came up with a scheme involving just `SeparableSum` and `VCAT`, but that seems fairly ugly. Is there a nicer way to just compute the sum of two functions on the same input?

Whatever the answer is, I am happy to try to add it to the demos or documentation if you think it would be helpful.

**Comment by nantonel** [comment]
Yes you're right, `SlicedSeparableSum` is missing this feature. Gradients haven't been implemented yet for calculus rules, but this should be possible and should be relatively easy for `SlicedSeparableSum`. 

In the meantime you can do this:
```julia
f = [SqrNormL2(10),SqrNormL2(10)]
grad = [zeros(10),zeros(20)] 
# allocate gradients
x = [randn(10),randn(20)]     
# points of evaluation
gradient!.(grad,f,x)               
# perform in place gradient
```

**Comment by lostella** [comment]
But I guess having `gradient!` for `SlicedSeparableSum` would simply offer an alternative unpleasent way of computing the gradient of the sum of f1+f2 :-)

The answer is: there is no idiomatic way as for now (that is, a simple `Sum` rule): this was not implemented as we initially focused on the `prox!` operation, which doesn't have a closed form expression for the sum of two functions in general. Since, however, we went more towards the direction of "first-order primitives" (prox as well as gradients) I guess having a `Sum` calculus rule wouldn't harm.

@currymj if you are interested in contributing, please feel free to open a PR: my suggestion is to create a `src/calculus/sum.jl` file defining the `Sum` type implementing function call and `gradient!` operation. This will be something along the lines of [https://github.com/kul-forbes/ProximalOperators.jl/blob/master/src/calculus/separableSum.jl]().

**Comment by currymj** [comment]
I'm giving that a shot right now. My own problem requires FBS with a sum of two quadratic forms (and only l1 regularization for g) so it will be directly useful.

**Comment by currymj** [comment]
Ok, PR open at #50 . I'm happy to make any needed modifications to make it more robust/useful.

**Comment by nantonel** [comment]
sorry I think I didn't get your question earlier! 😅 

Just to be clear: SlicedSeparableSum is meant only for non-overlapping indices to ensure the sum is separable.

I believe what you are looking for can be found in this [test of ProximalAlgorithms.jl](https://github.com/kul-forbes/ProximalAlgorithms.jl/blob/master/test/test_lasso_small_split_f.jl)

But that requires the usage VCAT from AbstractOperators.jl, so I think it's nice to provide an alternative by adding Sum here!

**Comment by currymj** [comment]
Yeah, `VCAT(Eye(d), Eye(d))` as the `Aq` argument and then `SeparableSum(f1, f2)` has the correct meaning. However I wasn't sure exactly how AbstractOperators.jl might handle that -- would it actually duplicate things?

i'm open to any and all criticism on #50, even if it's very vague!

**Comment by nantonel** [comment]
This is what roughly happens internally in `ProximalAlgorithms`  to calculate the gradient using `AbstractOperators`: 
given   
`Aq = VCAT(Eye(d),Eye(d))`
`fq = SeparableSum(f1,f2)` 
then first the iterate is multiplied  `Aqx = Aq*x = (x,x)`
then the gradient is applied `(gradf1,gradf2) = gradient(fq,Aqx)` 
and finally `grad = Aq'*(gradf1,gradf2)` sums them up. 

So you are right there is a duplication, but still all of this operations can be performed in-place.

**Comment by lostella** [comment]
Solved in #50/#55.
